### PR TITLE
Ignore invalid values for `page` parameter in searches

### LINKED ISSUE #13868: [production][] offset out of bounds (offset=93040, max_matches=15000) - SELECT *, (weight() + 100 * linked_count...
[See this exception on Errbit](http://errbit.opensuse.org/apps/5620ca2bdc71fa00b1000000/problems/63bcd5c784bf466764ed6160)
# offset out of bounds (offset=93040, max_matches=15000) - SELECT *, (weight() + 100 * linked_count + -1000 * links_to_other + 1000 * is_devel + 500 * (activity_index * POW( 2.3276, (updated_at - 1673319879) / 10000000))) as adjusted_weight FROM `package_core`, `project_core` WHERE MATCH(&#39;docker&#39;) AND `sphinx_deleted` = 0 AND `project_id` NOT IN (50315, 60546, 69200, 122039, 122041, 122042, 122043, 248287, 254068, 889915, 277778, 960219, 1087175, 1087178) ORDER BY `adjusted_weight` DESC LIMIT 93040, 20 OPTION cutoff=0, field_weights=(name=10, title=2, description=1), max_matches=15000, ranker=sph04; SHOW META #
### Occured ###
Jan 10  3:04am

### Similar ###
3

## Params ##
~~~
{}
~~~

## Session ##
~~~
{}
~~~

## Backtrace ##
~~~
70:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/connection/client.rb -> **rescue in perform**
66:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/connection/client.rb -> **perform**
51:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/connection/client.rb -> **check_and_perform**
36:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/connection/client.rb -> **query_all**
19:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/search/batch_inquirer.rb -> **block in results**
42:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/connection.rb -> **block in take**
138:  /GEM_ROOT/gems/innertube-1.1.0/lib/innertube.rb -> **take**
40:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/connection.rb -> **take**
18:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/search/batch_inquirer.rb -> **results**
11:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/middlewares/inquirer.rb -> **block in call**
206:  /GEM_ROOT/gems/activesupport-7.0.4/lib/active_support/notifications.rb -> **block in instrument**
24:  /GEM_ROOT/gems/activesupport-7.0.4/lib/active_support/notifications/instrumenter.rb -> **instrument**
206:  /GEM_ROOT/gems/activesupport-7.0.4/lib/active_support/notifications.rb -> **instrument**
5:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/logger.rb -> **log**
10:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/middlewares/inquirer.rb -> **call**
13:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/middlewares/geographer.rb -> **call**
16:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/middlewares/sphinxql.rb -> **call**
9:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/middlewares/valid_options.rb -> **call**
12:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/middlewares/stale_id_filter.rb -> **call**
31:  /GEM_ROOT/gems/middleware-0.1.0/lib/middleware/runner.rb -> **call**
102:  /GEM_ROOT/gems/middleware-0.1.0/lib/middleware/builder.rb -> **call**
95:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/search.rb -> **populate**
145:  /GEM_ROOT/gems/thinking-sphinx-5.5.1/lib/thinking_sphinx/search.rb -> **method_missing**
163:  app/controllers/webui/search_controller.rb -> **search**
7:  app/controllers/webui/search_controller.rb -> **index**
6:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_controller/metal/basic_implicit_render.rb -> **send_action**
215:  /GEM_ROOT/gems/actionpack-7.0.4/lib/abstract_controller/base.rb -> **process_action**
53:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_controller/metal/rendering.rb -> **process_action**
234:  /GEM_ROOT/gems/actionpack-7.0.4/lib/abstract_controller/callbacks.rb -> **block in process_action**
107:  /GEM_ROOT/gems/activesupport-7.0.4/lib/active_support/callbacks.rb -> **run_callbacks**
233:  /GEM_ROOT/gems/actionpack-7.0.4/lib/abstract_controller/callbacks.rb -> **process_action**
22:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_controller/metal/rescue.rb -> **process_action**
67:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_controller/metal/instrumentation.rb -> **block in process_action**
206:  /GEM_ROOT/gems/activesupport-7.0.4/lib/active_support/notifications.rb -> **block in instrument**
24:  /GEM_ROOT/gems/activesupport-7.0.4/lib/active_support/notifications/instrumenter.rb -> **instrument**
206:  /GEM_ROOT/gems/activesupport-7.0.4/lib/active_support/notifications.rb -> **instrument**
66:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_controller/metal/instrumentation.rb -> **process_action**
259:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_controller/metal/params_wrapper.rb -> **process_action**
27:  /GEM_ROOT/gems/activerecord-7.0.4/lib/active_record/railties/controller_runtime.rb -> **process_action**
151:  /GEM_ROOT/gems/actionpack-7.0.4/lib/abstract_controller/base.rb -> **process**
39:  /GEM_ROOT/gems/actionview-7.0.4/lib/action_view/rendering.rb -> **process**
188:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_controller/metal.rb -> **dispatch**
251:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_controller/metal.rb -> **dispatch**
49:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/routing/route_set.rb -> **dispatch**
32:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/routing/route_set.rb -> **serve**
18:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/routing/mapper.rb -> **block in &lt;class:Constraints&gt;**
48:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/routing/mapper.rb -> **serve**
50:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/journey/router.rb -> **block in serve**
32:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/journey/router.rb -> **each**
32:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/journey/router.rb -> **serve**
852:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/routing/route_set.rb -> **call**
72:  /GEM_ROOT/gems/flipper-0.26.0/lib/flipper/middleware/memoizer.rb -> **memoized_call**
37:  /GEM_ROOT/gems/flipper-0.26.0/lib/flipper/middleware/memoizer.rb -> **call**
15:  config/initializers/prefer_xml_over_html.rb -> **call**
44:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/deflater.rb -> **call**
15:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/tempfile_reaper.rb -> **call**
27:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/etag.rb -> **call**
27:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/conditional_get.rb -> **call**
12:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/head.rb -> **call**
38:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/http/permissions_policy.rb -> **call**
36:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/http/content_security_policy.rb -> **call**
266:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/session/abstract/id.rb -> **context**
260:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/session/abstract/id.rb -> **call**
696:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/middleware/cookies.rb -> **call**
27:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/middleware/callbacks.rb -> **block in call**
99:  /GEM_ROOT/gems/activesupport-7.0.4/lib/active_support/callbacks.rb -> **run_callbacks**
26:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/middleware/callbacks.rb -> **call**
32:  /GEM_ROOT/gems/airbrake-9.4.3/lib/airbrake/rack/middleware.rb -> **call!**
21:  /GEM_ROOT/gems/airbrake-9.4.3/lib/airbrake/rack/middleware.rb -> **call**
28:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/middleware/debug_exceptions.rb -> **call**
26:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/middleware/show_exceptions.rb -> **call**
18:  /GEM_ROOT/gems/lograge-0.12.0/lib/lograge/rails_ext/rack/logger.rb -> **call_app**
25:  /GEM_ROOT/gems/railties-7.0.4/lib/rails/rack/logger.rb -> **block in call**
99:  /GEM_ROOT/gems/activesupport-7.0.4/lib/active_support/tagged_logging.rb -> **block in tagged**
37:  /GEM_ROOT/gems/activesupport-7.0.4/lib/active_support/tagged_logging.rb -> **tagged**
99:  /GEM_ROOT/gems/activesupport-7.0.4/lib/active_support/tagged_logging.rb -> **tagged**
25:  /GEM_ROOT/gems/railties-7.0.4/lib/rails/rack/logger.rb -> **call**
93:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/middleware/remote_ip.rb -> **call**
93:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/middleware/remote_ip.rb -> **call**
19:  /GEM_ROOT/gems/request_store-1.5.1/lib/request_store/middleware.rb -> **call**
26:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/middleware/request_id.rb -> **call**
24:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/method_override.rb -> **call**
22:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/runtime.rb -> **call**
29:  /GEM_ROOT/gems/activesupport-7.0.4/lib/active_support/cache/strategy/local_cache_middleware.rb -> **call**
14:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/middleware/executor.rb -> **call**
18:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/lock.rb -> **call**
110:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/sendfile.rb -> **call**
131:  /GEM_ROOT/gems/actionpack-7.0.4/lib/action_dispatch/middleware/host_authorization.rb -> **call**
530:  /GEM_ROOT/gems/railties-7.0.4/lib/rails/engine.rb -> **call**
226:  /GEM_ROOT/gems/railties-7.0.4/lib/rails/railtie.rb -> **public_send**
226:  /GEM_ROOT/gems/railties-7.0.4/lib/rails/railtie.rb -> **method_missing**
74:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/urlmap.rb -> **block in call**
58:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/urlmap.rb -> **each**
58:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/urlmap.rb -> **call**
107:  /usr/lib64/ruby/gems/3.1.0/gems/passenger-6.0.14/src/ruby_supportlib/phusion_passenger/rack/thread_handler_extension.rb -> **process_request**
149:  /usr/lib64/ruby/gems/3.1.0/gems/passenger-6.0.14/src/ruby_supportlib/phusion_passenger/request_handler/thread_handler.rb -> **accept_and_process_next_request**
110:  /usr/lib64/ruby/gems/3.1.0/gems/passenger-6.0.14/src/ruby_supportlib/phusion_passenger/request_handler/thread_handler.rb -> **main_loop**
419:  /usr/lib64/ruby/gems/3.1.0/gems/passenger-6.0.14/src/ruby_supportlib/phusion_passenger/request_handler.rb -> **block (3 levels) in start_threads**
113:  /usr/lib64/ruby/gems/3.1.0/gems/passenger-6.0.14/src/ruby_supportlib/phusion_passenger/utils.rb -> **block in create_thread_and_abort_on_exception**
~~~

## Environment ##
| Key        | Value      |
|------------|------------|
| program_name | Passenger RubyApp: /srv/www/obs/api (production) |
| HTTP_USER_AGENT |  |

### PR DESCRIPTION
**Pull request by eduardoj**
Fixes #13868.

## For reviewers

- With an invalid value of the `page` parameter, the first page of the results is shown: https://obs-reviewlab.opensuse.org/eduardoj-fix_13868/search?search_text=another&page=some_random_text
- With a very big value of the `page` parameter, the first page of the results is shown: https://obs-reviewlab.opensuse.org/eduardoj-fix_13868/search?search_text=another&page=100000

### CODE REVIEW

File: src/api/app/controllers/webui/search_controller.rb [hunk 1]
```diff
@@ -161,9 +162,19 @@ def search
                                 issue_tracker_name: @search_tracker)
     @results = search.search(page: params[:page], per_page: @per_page)
     flash[:notice] = 'Your search did not return any results.' if @results.empty?
+  rescue ThinkingSphinx::OutOfBoundsError
+    flash[:error] = 'You are trying to access results out of bounds. Please adjust your "page" parameter accordingly.'
```
Comment (danidoni) [inline src/api/app/controllers/webui/search_controller.rb:166]: I would just discard the page param if the page parameter is not a number...

File: src/api/app/controllers/webui/search_controller.rb [hunk 2]
```diff
@@ -161,9 +162,19 @@ def search
...
   rescue ThinkingSphinx::SphinxError => e
     flash[:error] = "There has been an error performing the search. We're working to fix it. Please, try again later."
     Airbrake.notify(e)
     @results = []
   end
+
+  def validate_page
+    return unless params[:page].present? && params[:page] !~ /\A\d+\z/
+
+    flash[:error] = 'You must pass an integer for the "page" parameter.'
```
Comment (danidoni) [inline src/api/app/controllers/webui/search_controller.rb:177]: I would just discard the page param if the page parameter is not a number...

Comment (danidoni) [reply src/api/app/controllers/webui/search_controller.rb:177]: Sorry, what I meant was that you could just discard the error message and just redirect back without telling the user anything.

Comment (eduardoj) [reply src/api/app/controllers/webui/search_controller.rb:177]: I prefer to be verbose and tell the user what has gone wrong. Do you find a reason to hide those details to the user?

**Review by danidoni — APPROVED** [review]

File: src/api/app/controllers/webui/search_controller.rb (hunk 2 above)
Comment (danidoni) [reply src/api/app/controllers/webui/search_controller.rb]: LGTM

**Pull request merged**